ntoskrnl/io/iomgr/driver.c

```diff
@@ -2181,4 +2194,233 @@ NtUnloadDriver(IN PUNICODE_STRING DriverServiceName)
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ExFreePool(SubKeyName.Buffer);
+    ZwClose(RootKeyHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"NextInstance");
+    RegValue = 1;
+    Status = ZwSetValueKey(KeyHandle,
+                           &KeyString,
+                           0,
+                           REG_DWORD,
+                           &RegValue,
+                           sizeof(RegValue));
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"000");
+    Status = IopCreateRegistryKeyEx(&NodeHandle,
+                                    KeyHandle,
+                                    &KeyString,
+                                    KEY_ALL_ACCESS,
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ZwClose(KeyHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    Status = RtlDuplicateUnicodeString(RTL_DUPLICATE_UNICODE_STRING_NULL_TERMINATE,
+                                       ServiceName,
+                                       &ServiceName1);
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"Service");
+    Status = ZwSetValueKey(NodeHandle,
+                           &KeyString,
+                           0,
+                           REG_SZ,
+                           ServiceName->Buffer,
+                           ServiceName->Length + sizeof(UNICODE_NULL));
+    if (!NT_SUCCESS(Status))
+    {
+        ExFreePool(ServiceName1.Buffer);
+        goto Quit;
+    }
+
+    /* Try to get the DisplayName value from the service key */
+    /* and use it as the DeviceDesc entry. */
+    DisplayName.Length = 0;
+    DisplayName.Buffer = NULL;
+
+    Status = IopGetRegistryValue(ServiceHandle, L"DisplayName", &KeyValueInfo);
+    if (NT_SUCCESS(Status))
+    {
+        if ((KeyValueInfo->Type == REG_SZ) &&
+            (KeyValueInfo->DataLength > 0))
+        {
+            DisplayName.Length = KeyValueInfo->DataLength - sizeof(UNICODE_NULL);
+            DisplayName.MaximumLength = KeyValueInfo->DataLength;
+            DisplayName.Buffer = ExAllocatePoolWithTag(PagedPool, DisplayName.MaximumLength, TAG_IO);
+            if (DisplayName.Buffer != NULL)
+            {
+                RtlMoveMemory(DisplayName.Buffer,
+                              (PVOID)((ULONG_PTR)KeyValueInfo + KeyValueInfo->DataOffset),
+                              DisplayName.Length);
+                DisplayName.Buffer[DisplayName.Length / sizeof(WCHAR)] = UNICODE_NULL;
+            }
+        }
+        ExFreePool(KeyValueInfo);
+    }
+
+    /* If there is no DisplayName we just use the ServiceName. */
+    if (DisplayName.Length == 0)
```

Comment:
It might now be better to instead do the following check:
```suggestion
    if (DisplayName.Buffer == NULL)
```